# Лекция 01. Обзор Python

Переводим знакомые конструкции на Python. Две задачи проведут нас через основную часть: проверка длины имени в игре и чтение книги по страницам.

## Давайте познакомимся

Меня зовут **Сергей Нечаев**, я преподаватель курса «Углублённый Python». Будем учиться читать код, писать небольшие проверяемые функции, а затем собирать из них приложения. Опыт программирования на другом языке пригодится; знание Python заранее не требуется.

- У нас 14 лекций и 14 семинаров. Материалы занятия: `lecture.ipynb`, `seminar.ipynb`, `tasks.md`.
- На лекции предсказываем результат, запускаем код и объясняем расхождения. На семинаре пишем сами; дома решаем 3 Easy и 2 Medium, иногда есть необязательная Hard-задача.
- Домашние работы нужно защищать. Контрольная и экзамен проходят очно и письменно. Проект выполняется в команде из 2–4 человек, и каждый участник должен понимать работу группы.
- Оценка за курс при сдаче экзамена: **0,6 × семинары + 0,1 × контрольная + 0,3 × экзамен**. Семинары: **0,5 × домашние работы + 0,4 × проект + бонус семинариста от 0 до 1**.
- Если семинары и контрольная **оба больше 6**, доступен автомат: **min(семинары, контрольная, 8)**. Сдача экзамена может повысить или понизить результат.

Полные правила сохранены в справочной части этого ноутбука и в [программе курса](../COURSE_PLAN.md). Сроки и штрафы публикуются в [DEADLINES.md](../DEADLINES.md). К вопросам о правилах можно вернуться после основной части.

## Цели

К концу занятия вы сможете по небольшому словесному условию написать функцию на Python и проверить обычный случай и границы.

Для этого разберём значения и типы, условия, функции с `return`, циклы `for` и `while`, `break` и `continue`. Научимся читать базовые аннотации типов и находить место простой ошибки. Познакомимся с Zen of Python и основными правилами оформления из PEP 8.

## Перед началом

Вы уже знакомы с переменными, условиями, циклами и функциями на уровне понятий. Сегодня выясняем, как они записываются и работают в Python.

Используем Python 3.14 и Jupyter; инструкция запуска есть в [README курса](../README.md). Ячейка выполняется сочетанием **Shift+Enter**. В ноутбуке есть отдельные ячейки с текстом и с кодом.

### Zen of Python: как выбирать понятное решение

**PEP** (Python Enhancement Proposal) — документ сообщества о развитии Python, рабочих правилах или рекомендациях. [PEP 20 — The Zen of Python](https://peps.python.org/pep-0020/) содержит короткие принципы, сформулированные Тимом Петерсом. Они помогают обсуждать решения и читаемость кода.

Полный текст на английском можно вывести прямо в Python. `import` подключает модуль; модуль `this` при первом импорте в текущем сеансе Python печатает Zen of Python.

In [ ]:
import this

Для начала возьмём четыре принципа. Примеры ниже подробнее разберём по ходу лекции:

- **Readability counts — читаемость имеет значение.** По имени `pages_left` легче понять смысл значения, чем по `p`. К этому коду позже вернётся и его автор.
- **Explicit is better than implicit — явное лучше неявного.** В `int(raw_pages)` видно намерение получить число из текста. Проверка `pages_read is None` точно выражает отсутствие данных, когда `0` — допустимое число страниц.
- **Simple is better than complex — простое лучше сложного.** Условие `3 <= len(name) <= 12` прямо повторяет правило о допустимой длине. Решение проще проверить, если его шаги легко объяснить.
- **Although practicality beats purity — практичность важнее безупречного следования принципу.** Выбираем запись с учётом задачи и читателя. Если сложное выражение трудно разобрать, можно дать промежуточному результату имя.

Эти принципы пригодятся при обсуждении нескольких работающих решений: какое проще прочитать, проверить и изменить?

### PEP 8: немного об оформлении

[PEP 8](https://peps.python.org/pep-0008/) — рекомендации по оформлению Python-кода. Единообразная запись помогает быстрее замечать смысл и ошибки. Для первых задач достаточно нескольких правил:

- **Четыре пробела на уровень отступа.** Настройте редактор на вставку пробелов; не смешивайте их с табуляцией.
- **Имена переменных и функций — `snake_case`:** `pages_left`, `nickname_status`. Подбирайте имена по смыслу.
- **Пробелы вокруг `=` при присваивании и вокруг сравнений, после запятых.** В аннотациях пишем `name: str` и ` -> bool`; внутри скобок лишние пробелы не нужны.
- **Тело условия, цикла или функции — с новой строки.** Несколько инструкций через `;` затрудняют чтение.
- **В `.py`-файле между определениями функций верхнего уровня — две пустые строки.** Внутри функции пустой строкой можно отделить логический шаг.
- **Ориентир для длины строки — 79 символов.** Длинные выражения можно переносить внутри скобок. В проекте учитывайте его принятые правила оформления.

Отступы определяют блоки кода и влияют на выполнение. Четыре пробела — рекомендация по их ширине; многие нарушения стиля сами по себе не мешают запуску.

Например, эта функция работает, но читать её неудобно:

```python
def isValidName(name:str)->bool:return 3<=len(name)<=12
```

Сравните с записью ниже. Изменились имя и оформление; правило проверки осталось тем же.

In [ ]:
def is_valid_name(name: str) -> bool:
    return 3 <= len(name) <= 12


print(is_valid_name("кот"))
print(is_valid_name("я"))

**Быстрая проверка перед объяснением:** предскажите три результата следующей ячейки. Здесь `print` выводит значение, кавычки обозначают текст, а `<=` — сравнение «меньше или равно». Если запись незнакома, достаточно объяснить свою догадку.

In [1]:
print(10 + 20)
print("10" + "20")
print(1 <= 3 <= 5)

30
1020
True


Получились число `30`, строка `1020` и логическое значение `True`. Один знак `+` складывает числа или соединяет строки — это зависит от типов значений. Цепочка сравнений проверяет обе границы.

Если всё предсказали верно, дальше ищите отличия от знакомого языка. Если ответы разошлись, последующие примеры помогут восстановить правила.

## Как выполняется код

В `.py`-файле инструкции выполняются сверху вниз; условия, циклы и вызовы функций меняют порядок. Выражение вычисляет значение, например `120 - 35`; присваивание `pages_left = 120 - 35` связывает имя с результатом. `=` — присваивание, `==` — проверка равенства.

В Jupyter выполненные ячейки используют один процесс Python — **kernel**. Имена сохраняются до его перезапуска. Номер возле ячейки показывает порядок её фактического запуска. **Restart Kernel and Run All Cells** позволяет проверить ноутбук с чистого состояния.

Предскажите, что напечатает пример. Изменится ли `pages_left` само, когда мы переприсвоим `pages_read`?

In [2]:
total_pages = 120
pages_read = 35
pages_left = total_pages - pages_read

pages_read = 40
print(pages_left)

pages_left = total_pages - pages_read
print(pages_left)

85
80


Сначала выводится `85`, затем `80`. Присваивание сохранило результат вычисления, а не формулу, которая автоматически пересчитывается при изменении других имён.

На протяжении лекции задаём три вопроса: **какое значение получилось, какой у него тип и какая инструкция выполнится следующей?**

## Значения, типы и преобразования

| Пример | Тип | Что хранит |
| --- | --- | --- |
| `120` | `int` | целое число |
| `2.5` | `float` | число с плавающей точкой |
| `"120"` | `str` | текст |
| `True`, `False` | `bool` | логическое значение |
| `None` | `NoneType` | специальное значение для отсутствующих данных |

`type(value)` показывает тип, `int("120")` преобразует текст в целое число, `str(120)` — число в текст. Преобразование должно иметь смысл: строку `"книга"` нельзя превратить в целое число через `int`.

Предскажите значения и типы результатов. В Python тип есть у значения; имя можно связать с другим значением.

In [3]:
raw_pages = "120"
total_pages = int(raw_pages)

print(raw_pages + "20")
print(total_pages + 20)
print(type(raw_pages), type(total_pages))

12020
140
<class 'str'> <class 'int'>


## Вычисление и вывод

Обычные `+`, `-`, `*` работают с числами. `/` выполняет деление и возвращает `float`, `//` — деление с округлением вниз, `%` — остаток. Скобки делают порядок вычислений явным.

Книгу из 125 страниц читают по 20 страниц за день. Сколько получится полных порций по 20 страниц и сколько страниц останется?

Строка с префиксом `f` подставляет значения из фигурных скобок. Запись `{value:.2f}` выводит число с двумя знаками после точки; она меняет представление, а не само число.

In [4]:
total_pages = 125
pages_per_day = 20
full_days = total_pages // pages_per_day
remaining_pages = total_pages % pages_per_day

print(f"Полных порций: {full_days}; осталось страниц: {remaining_pages}")
print(f"Частное: {total_pages / pages_per_day:.2f}")

Полных порций: 6; осталось страниц: 5
Частное: 6.25


## Условия: длина имени в игре

Пусть по правилам игры имя должно содержать **от 3 до 12 символов включительно**. Нам нужны три ответа: `too short`, `too long` и `ok`. Длину строки возвращает `len(name)`.

`if` проверяет первое условие, `elif` — следующее, если предыдущие ложны, `else` выполняется, если ни одно не подошло. После условия ставится двоеточие, тело ветви записывается с отступом в четыре пробела. В одной цепочке выполняется не более одной ветви.

До запуска назовите ответ для `"кот"`. Затем проверьте строки длиной 2, 3, 12 и 13. Почему для проверки недостаточно одного имени средней длины?

In [ ]:
name = "кот"
length = len(name)

if length < 3:
    status = "too short"
elif length > 12:
    status = "too long"
else:
    status = "ok"

print(status)

## Несколько условий и truthiness

Для ответа «подходит или нет» достаточно `3 <= len(name) <= 12`. Это та же проверка, что `len(name) >= 3 and len(name) <= 12`.

`and` требует истинности обоих условий, `or` — хотя бы одного, `not` меняет логическое значение на противоположное. В логическом контексте ложны `False`, `None`, числовой ноль, пустые строки и контейнеры; большинство остальных значений истинны. `bool(value)` позволяет явно увидеть это значение.

`if name:` проверяет непустоту, но не допустимую длину. `if pages_read is None:` проверяет отсутствие данных: **0 прочитанных страниц и неизвестное число страниц — разные случаи**.

In [ ]:
name = "я"
print(bool(name))
print(3 <= len(name) <= 12)

pages_read = 0
if pages_read is None:
    print("Число прочитанных страниц неизвестно")
else:
    print(f"Прочитано страниц: {pages_read}")

**Короткое замыкание:** `and` и `or` не вычисляют правую часть, если левая уже определила исход проверки. Объясните, почему следующий пример не делит на ноль. Потом замените `days` на `3`.

In [ ]:
days = 0
pages_read = 60
fast_reading = days > 0 and pages_read / days >= 20
print(fast_reading)

## От проверки к функции

Чтобы проверять разные имена, вынесем правило в функцию. `def` создаёт функцию; её тело выполняется при вызове. `name` в определении — **параметр**, строка `"кот"` при вызове — **аргумент**.

`return` немедленно завершает функцию и передаёт результат вызывающему коду. Если выполнение дошло до конца тела без `return`, результатом будет `None`.

`name: str` и `-> str` — аннотации ожидаемых типов аргумента и результата. Они помогают читать код, но сам интерпретатор не проверяет соответствие им автоматически.

Перед запуском найдите строку, на которой закончится каждый вызов.

In [ ]:
def nickname_status(name: str) -> str:
    length = len(name)
    if length < 3:
        return "too short"
    if length > 12:
        return "too long"
    return "ok"


print(nickname_status("я"))
print(nickname_status("кот"))
print(nickname_status("оченьдлинноеимя"))

## Проверяем обычный случай и границы

Функция получает данные через параметры и возвращает результат. Теперь её можно проверить без ручной замены имён в программе.

`assert condition` ничего не выводит, если условие истинно, и сообщает об `AssertionError`, если оно ложно. Начнём с обычного случая, затем проверим пустую строку и значения с обеих сторон границ.

Запись `"a" * 12` повторяет строку 12 раз. Какой `assert` обнаружит ошибку, если в функции заменить `< 3` на `<= 3`?

In [ ]:
assert nickname_status("игрок") == "ok"
assert nickname_status("") == "too short"
assert nickname_status("aa") == "too short"
assert nickname_status("aaa") == "ok"
assert nickname_status("a" * 12) == "ok"
assert nickname_status("a" * 13) == "too long"

## Почему `print` не заменяет `return`

`print` выводит значение для человека. `return` передаёт результат программе, которая вызвала функцию. Предскажите весь вывод и значение переменной `result`.

In [ ]:
def show_length(text: str):
    print(len(text))


result = show_length("кот")
print(result)

Первый вызов напечатал `3`, а второй — `None`: в `show_length` нет `return`. Чтобы использовать длину в дальнейших вычислениях, функция должна возвращать `len(text)`.

Параметры и имена, созданные внутри функции, локальны. Например, `length` внутри `nickname_status` не меняет одноимённую переменную снаружи. Проверим это отдельным вызовом.

In [ ]:
length = 100
result = nickname_status("кот")
print(result, length)

## `for`: читаем заданное число дней

За первый день прочитали 10 страниц, за второй — 20, за третий — 30. Сколько всего прочитано? Перед запуском проследите значения `day` и `pages_read` по шагам.

`for` перебирает значения. `range(1, 4)` даёт `1`, `2`, `3`: правая граница не включается. Перед циклом создаём накопитель, внутри обновляем его. `pages_read += today` означает `pages_read = pages_read + today`.

In [ ]:
pages_read = 0

for day in range(1, 4):
    today = day * 10
    pages_read += today
    print(day, today, pages_read)

print(f"Всего страниц: {pages_read}")

## `while`: читаем до конца книги

Теперь книга содержит 65 страниц, а за день можно прочитать до 20. Нужно определить число дней; последний день может быть неполным.

`while` проверяет условие перед каждой итерацией. В теле должно меняться состояние, от которого зависит продолжение цикла. Здесь будем уменьшать число оставшихся страниц, не опуская его ниже нуля: `min(a, b)` возвращает меньшее из двух значений.

До запуска назовите ответ. Затем проверьте книги из 0, 20 и 21 страницы. Почему при 0 страниц тело цикла не выполняется ни разу?

In [ ]:
pages_left = 65
pages_per_day = 20
days = 0

while pages_left > 0:
    today = min(pages_left, pages_per_day)
    pages_left -= today
    days += 1

print(days)

## `break` и `continue`

`break` завершает ближайший цикл. `continue` пропускает оставшуюся часть текущей итерации и переходит к следующей.

В следующем примере пропустим дни, кратные трём, и остановим чтение, когда книга закончится. На остальных днях читаем до 20 страниц. Операция `day % 3 == 0` проверяет кратность трём.

Предскажите, какие дни будут напечатаны. Пальцем проследите переходы: куда уводит `continue`, а куда — `break`?

In [ ]:
pages_left = 65

for day in range(1, 8):
    if day % 3 == 0:
        continue
    pages_left -= min(pages_left, 20)
    print(day, pages_left)
    if pages_left == 0:
        break

## Как прочитать ошибку

**Намеренно ошибочный пример.** Он показан текстом, чтобы общий запуск ноутбука не останавливался. Объясните причину, затем запустите исправленный вариант в следующей ячейке.

```python
pages = "120"
print(pages + 20)
```

При запуске такого `.py`-файла конец сообщения выглядит так:

```text
  File "example.py", line 2, in <module>
    print(pages + 20)
TypeError: can only concatenate str (not "int") to str
```

Читаем снизу вверх: **тип ошибки** `TypeError`, **причина** — попытка соединить строку с числом, **место** — строка 2. В Jupyter вместо имени файла будет ссылка на ячейку. Если в сообщении несколько вызовов, начинайте с последнего участка вашего кода.

Исправление зависит от задачи: для числового сложения нужно `int(pages) + 20`, для соединения текста — `pages + "20"`.

In [ ]:
pages = "120"
result = int(pages) + 20
assert result == 140
print(result)

## Неожиданно, но по правилам

**Сначала предскажите, затем запустите, потом прочитайте объяснение ниже.**

1. Какие логические значения получатся из строк `"False"` и `"0"`?
2. Чему равны `-7 // 2` и `-7 % 2`? Сохраняется ли равенство `a == (a // b) * b + a % b`?
3. Что произойдёт, если в функцию с аннотацией `int` передать строку?

In [5]:
print(bool("False"), bool("0"))

quotient = -7 // 2
remainder = -7 % 2
print(quotient, remainder)
print(quotient * 2 + remainder)


def double(value: int) -> int:
    return value * 2


print(double(3))
print(double("ha"))

True True
-4 1
-7
6
haha


**Разбор после запуска.**

1. Обе строки непустые, поэтому получаем `True`, `True`. Их содержимое не превращается в логическую команду.
2. Получаем `-4` и `1`. Деление `//` округляет вниз, в сторону меньшего числа; `-4 * 2 + 1 == -7`.
3. Получаем `6` и `haha`. Аннотация не запретила вызов со строкой, а умножение строки на целое число повторило её.

## Самопроверка

1. Почему присваивание `pages_left = total_pages - pages_read` не создаёт автоматически пересчитываемую формулу?
2. Чем отличаются `120` и `"120"`? Как получить число из строки?
3. Чем различаются `if pages_read:` и `if pages_read is None:` при значении `0`?
4. Какие длины строк проверяют обе границы правила «от 3 до 12 включительно»?
5. Почему после `return` следующая инструкция функции не выполнится? Что будет без `return`?
6. Что перебирает `for day in range(1, 4)`? Когда лучше выбрать `while`?
7. Куда передают управление `break` и `continue`?
8. Что проверяют аннотации типов во время обычного запуска программы?
9. Какой принцип Zen of Python помогает выбирать между работающими решениями? Назовите два правила PEP 8, которые делают код понятнее.

Напишите функцию `is_short(text: str) -> bool`, которая возвращает `True`, если длина строки меньше 5. Добавьте проверки пустой строки и строк длиной 4 и 5. Объясните каждый результат.

## Итоги

Мы превратили правило о длине строки в функцию и проверили её на границах. Затем проследили накопление результата в `for`, завершение процесса в `while` и прочитали сообщение об ошибке типов. Обсудили читаемость через принципы Zen of Python и применили базовые правила PEP 8 к знакомой проверке длины.

Рабочая последовательность: **сформулировать правило → записать функцию → предсказать результат → проверить обычный случай и границы**.

Дальше — [семинар](seminar.ipynb) и [пять домашних задач](tasks.md). На занятии 2 подробнее разберём строки, списки и файлы.

# Справочно: устройство курса

Полные организационные правила. В основной части преподаватель кратко объясняет механику оценивания; эти разделы остаются для самостоятельного чтения и вопросов.

## Как устроен курс

| Компонент | Что будет | Зачем |
| --- | ---: | --- |
| Лекции | 14 | вводят модель языка, алгоритмы, структуры данных и инженерные темы |
| Практические семинары | 10 | переводят идеи лекции в работающий код |
| Справочные семинары | 3 | дают примеры для проектов без обязательной практической сдачи |
| Защита проекта | 1 семинар | завершает курс демонстрацией групповой работы |
| Домашние работы | 13 | закрепляют каждую лекцию, кроме последней |
| Контрольная работа | 1 | очно проверяет базовые конструкции и алгоритмические темы |
| Экзамен | 1 | очно проверяет итоговое владение материалом |
| Групповой проект | 1 | связывает алгоритмы, качество кода и разработку приложения |

Каждое занятие живёт в каталоге `lesson??`: там находятся `lecture.ipynb`, `seminar.ipynb`, `tasks.md` и краткий `README.md`.

## Как складывается оценка

Все оценки выставляются по **десятибалльной шкале**. Для студента, который сдаёт экзамен, итоговая оценка рассчитывается так:

$$
O_{\text{курс}} = 0{,}6 \cdot O_{\text{семинары}} + 0{,}1 \cdot O_{\text{контрольная}} + 0{,}3 \cdot O_{\text{экзамен}}.
$$

Оценка за семинары включает домашние работы, групповой проект и работу на занятиях:

$$
O_{\text{семинары}} = 0{,}5 \cdot O_{\text{домашние работы}} + 0{,}4 \cdot O_{\text{проект}} + B_{\text{семинариста}},
$$

где $B_{\text{семинариста}}$ — бонус от 0 до 1 балла за содержательную работу на семинарах. Его выставляет преподаватель соответствующей семинарской группы. Таким образом, домашние работы дают до 5 баллов семинарской оценки, проект — до 4, работа на занятиях — ещё до 1.

## Как проходит проверка

**Контрольная работа и экзамен проходят очно и письменно.** Основная часть состоит из относительно простых заданий по материалам курса: требуется не угадывать редкий трюк, а уверенно применять разобранные конструкции и объяснять ход решения.

Домашние работы остаются заочными, но сдача кода сама по себе не завершает проверку. Работу нужно **защитить** у ассистента или семинариста. Проверяющий выборочно выбирает задачи и может попросить:

- объяснить идею и отдельную строку решения;
- разобрать граничный случай;
- оценить сложность;
- предсказать поведение на изменённом входе;
- внести небольшое изменение в код.

Защита подтверждает, что студент понимает сданное решение. Возможность пользоваться инструментами при подготовке домашней работы не заменяет этого понимания.

## Что означают оценки 8 и 10

Курс непрофильный, поэтому стабильная работа по программе должна позволять получить **8** без сверхусилий. Оценки **9–10** требуют более уверенного и самостоятельного владения материалом.

Практический ориентир для верхней части шкалы: студент уверенно решает незнакомую задачу уровня **LeetCode Medium** по объявленным алгоритмическим темам, применяет разобранные конструкции Python, объясняет корректность решения и оценивает его время и память. Речь не о заученном решении конкретной задачи, а о способности самостоятельно перенести изученный метод на новый вход.

## Автомат и экзамен

Если одновременно $O_{\text{семинары}} > 6$ и $O_{\text{контрольная}} > 6$, студент может не сдавать экзамен и получить автомат:

$$
O_{\text{автомат}} = \min(O_{\text{семинары}}, O_{\text{контрольная}}, 8).
$$

Автомат не может быть выше 8. Если студент вместо автомата сдаёт экзамен, применяется обычная итоговая формула с весом экзамена 0,3. В этом случае экзамен может как повысить, так и понизить результат относительно доступного автомата.

Правила пересдач будут согласованы и опубликованы отдельно.

## Рабочий ритм

1. На лекции сначала **предсказываем** результат примера, затем запускаем код и объясняем расхождение.
2. На практическом семинаре решаем небольшие задачи и обсуждаем границы, ошибки и читаемость.
3. После занятия выполняем домашнюю работу: обычно **3 Easy + 2 Medium**, иногда есть необязательная Hard-задача.
4. Решение оформляем в требуемом формате и защищаем у ассистента или семинариста.

Семинар не обязан повторять задачи и примеры лекции. Семинарист может разбирать те же темы на других задачах, если они помогают освоить заявленные результаты обучения занятия.

Рекомендуемая траектория: 3 Easy — базовый уровень; 3 Easy + 1 Medium — хороший; весь основной набор — высокий. Hard-задачи остаются бонусным усложнением. Актуальные сроки и правила штрафов публикуются только в [`DEADLINES.md`](../DEADLINES.md); сейчас даты ещё не объявлены.

## Групповой проект

Группа из **2–4 человек** выбирает предметную задачу из предложенного списка и постепенно строит приложение. Проект формирует 40% оценки за семинары. Оценивается не количество модных технологий, а соответствие решения задаче, качество реализации и способность объяснить выбор.

Проект подразумевает вовлечённость и значимый вклад каждого участника. На защите каждый должен понимать задачу, основной сценарий, архитектуру, ключевые решения и собственный вклад. Если хотя бы один участник не может объяснить, что сделала группа, проект получает **незачёт для всей команды**, то есть $O_{\text{проект}} = 0$.

Контрольные точки распределены по курсу:

- после занятия 4 — тема, постановка задачи и репозиторий;
- после занятия 8 — предметное ядро и первоначальная архитектура;
- после занятия 10 — структура приложения, типизация, тесты и логирование;
- после занятия 11 — алгоритмическая часть и анализ сложности;
- после занятия 13 — внешние интерфейсы и хранение данных, если они нужны;
- занятие 14 — демонстрация и защита.

## Программа может учитывать запрос потока

Если у потока возникнет массовый запрос на анализ данных, одну из лекций заменим занятием по **NumPy, pandas и связанным инструментам**. Какую именно тему заменяет такой блок и в каком объёме он проводится, преподаватели объявят заранее.

## Как получать помощь

Ошибка — нормальная часть работы, но для содержательного вопроса нужен контекст. Покажите:

- что вы хотели получить;
- минимальный фрагмент кода, который воспроизводит проблему;
- входные данные;
- фактический и ожидаемый результат;
- полный traceback, если возникло исключение.

Фраза «не работает» почти не оставляет зацепок. Формулировка «на входе `0` ожидал `0`, получил `None`; вот функция и traceback» обычно позволяет быстро найти причину.

# Справочно: короткие памятки по Python

Эти сведения можно посмотреть при решении задачи. Учить таблицы наизусть перед семинаром не требуется.

## Имена, выражения и операции

- Python различает регистр: `name` и `Name` — разные имена. Имена могут содержать буквы, цифры и `_`, но не начинаются с цифры и не совпадают с ключевыми словами.
- `#` начинает комментарий. Число `10_000` равно `10000`: подчёркивание помогает читать разряды.
- `+`, `-`, `*`, `/`, `//`, `%`, `**` — сложение, вычитание, умножение, деление, деление с округлением вниз, остаток и степень.
- `==`, `!=`, `<`, `<=`, `>`, `>=` — сравнения. В сложном выражении порядок лучше обозначать скобками.
- `range(stop)` начинает с 0, `range(start, stop)` — с `start`; третий аргумент задаёт шаг. Правая граница всегда исключена. Например, `range(0, 7, 2)` даёт 0, 2, 4, 6.
- `input()` возвращает строку. В задачах этого занятия данные приходят через параметры, поэтому `input()` не нужен; подробнее разберём ввод на занятии 2.

## Именованные аргументы и значения по умолчанию

Аргументы можно передать по позиции или по имени параметра. Значение по умолчанию используется, если аргумент не передан.

Сигнатура описывает параметры функции; аннотации дополнительно указывают ожидаемые типы параметров и результата. Следующая функция по умолчанию проверяет длину от 3, а при явном аргументе `minimum` — от указанной границы.

In [ ]:
def is_long_enough(text: str, minimum: int = 3) -> bool:
    return len(text) >= minimum


print(is_long_enough("кот"))
print(is_long_enough("кот", 5))
print(is_long_enough(minimum=2, text="ёж"))

# Дополнительно: изменяемое значение по умолчанию

Списки подробно появятся на занятии 2, общие ссылки и изменяемость — на занятии 5.

Для чтения примера достаточно трёх пояснений: `[]` создаёт список, `append` добавляет элемент в существующий список, `a is b` проверяет, один ли это объект.

Предскажите содержимое `first` и `second` после двух вызовов. Мы печатаем оба результата только после второго вызова.

In [ ]:
def add_note(note: str, notes=[]):
    notes.append(note)
    return notes


first = add_note("прочитать главу")
second = add_note("решить задачу")
print(first)
print(second)
print(first is second)

**Разбор после запуска.** Значение по умолчанию создаётся один раз при выполнении `def`. Оба вызова без второго аргумента получают один список, поэтому оба имени показывают две заметки, а `first is second` даёт `True`.

Чтобы получать новый список на каждый вызов, используют `None` и создают список внутри функции. Правило относится к изменяемым значениям по умолчанию; передача аргумента по имени сама по себе его не меняет.

In [ ]:
def add_note_safely(note: str, notes=None):
    if notes is None:
        notes = []
    notes.append(note)
    return notes


safe_first = add_note_safely("прочитать главу")
safe_second = add_note_safely("решить задачу")
print(safe_first)
print(safe_second)
print(safe_first is safe_second)